messaging/messaging.go

```diff
@@ -0,0 +1,454 @@
+	ActionLocKey string   `json:"action-loc-key,omitempty"`
+	LaunchImage  string   `json:"launch-image,omitempty"`
+}
+
+// ErrorInfo is a topic management error.
+type ErrorInfo struct {
+	Index  int
+	Reason string
+}
+
+// TopicManagementResponse is the result produced by topic management operations.
+//
+// TopicManagementResponse provides an overview of how many input tokens were successfully handled,
+// and how many failed. In case of failures, the Errors list provides specific details concerning
+// each error.
+type TopicManagementResponse struct {
+	SuccessCount int
+	FailureCount int
+	Errors       []*ErrorInfo
+}
+
+func newTopicManagementResponse(resp *iidResponse) *TopicManagementResponse {
+	tmr := &TopicManagementResponse{}
+	for idx, res := range resp.Results {
+		if len(res) == 0 {
+			tmr.SuccessCount++
+		} else {
+			tmr.FailureCount++
+			code, _ := res["error"].(string)
+			reason := iidErrorCodes[code]
+			if reason == "" {
+				reason = "unknown-error"
+			}
+			tmr.Errors = append(tmr.Errors, &ErrorInfo{
+				Index:  idx,
+				Reason: reason,
+			})
+		}
+	}
+	return tmr
+}
+
+// NewClient creates a new instance of the Firebase Cloud Messaging Client.
+//
+// This function can only be invoked from within the SDK. Client applications should access the
+// the messaging service through firebase.App.
+func NewClient(ctx context.Context, c *internal.MessagingConfig) (*Client, error) {
+	if c.ProjectID == "" {
+		return nil, errors.New("project id is required to access firebase cloud messaging client")
+	}
+
+	hc, _, err := transport.NewHTTPClient(ctx, c.Opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		fcmEndpoint: messagingEndpoint,
+		iidEndpoint: iidEndpoint,
+		client:      &internal.HTTPClient{Client: hc},
+		project:     c.ProjectID,
+		version:     "Go/Admin/" + c.Version,
+	}, nil
+}
+
+// Send sends a Message to Firebase Cloud Messaging.
+//
+// The Message must specify exactly one of Token, Topic and Condition fields. FCM will
+// customize the message for each target platform based on the parameters specified in the
+// Message.
+func (c *Client) Send(ctx context.Context, message *Message) (string, error) {
+	payload := &fcmRequest{
+		Message: message,
+	}
+	return c.makeSendRequest(ctx, payload)
+}
+
+// SendDryRun sends a Message to Firebase Cloud Messaging in the dry run (validation only) mode.
+//
+// This function does not actually deliver the message to target devices. Instead, it performs all
+// the SDK-level and backend validations on the message, and emulates the send operation.
+func (c *Client) SendDryRun(ctx context.Context, message *Message) (string, error) {
+	payload := &fcmRequest{
+		ValidateOnly: true,
+		Message:      message,
+	}
+	return c.makeSendRequest(ctx, payload)
+}
+
+// SubscribeToTopic subscribes a list of registration tokens to a topic.
+//
+// The tokens list must not be empty, with at most 1000 tokens.
```

Comment:
nit: s/, with at most/ and must have at most/
(or something like that...)

Reply:
Done